core/capabilities/compute/compute.go

```diff
@@ -0,0 +1,202 @@
+	return nil
+}
+
+func (c *Compute) UnregisterFromWorkflow(ctx context.Context, request capabilities.UnregisterFromWorkflowRequest) error {
+	return nil
+}
+
+func generateID(binary []byte) string {
+	id := sha256.Sum256(binary)
+	return fmt.Sprintf("%x", id)
+}
+
+func (c *Compute) Execute(ctx context.Context, request capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	binary, err := c.popBytesValue(request.Config, binaryKey)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	config, err := c.popBytesValue(request.Config, configKey)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	id := generateID(binary)
+
+	m, err := c.modules.get(id)
+	if err != nil {
+		mod, err := c.initModule(id, binary, request.Metadata.WorkflowID, request.Metadata.ReferenceID)
+		if err != nil {
+			return capabilities.CapabilityResponse{}, err
+		}
+
+		m = mod
+	}
+
+	return c.executeWithModule(m.module, config, request)
+}
+
+func (c *Compute) initModule(id string, binary []byte, workflowID, referenceID string) (*module, error) {
+	initStart := time.Now()
+	mod, err := host.NewModule(&host.ModuleConfig{Logger: c.log}, binary)
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate WASM module: %w", err)
+	}
+
+	mod.Start()
+
+	initDuration := time.Now().Sub(initStart)
+	computeWASMInit.WithLabelValues(workflowID, referenceID).Observe(float64(initDuration))
+
+	m := &module{module: mod}
+	c.modules.add(id, m)
+	return m, nil
+}
+
+func (c *Compute) popBytesValue(m *values.Map, key string) ([]byte, error) {
+	v, ok := m.Underlying[key]
+	if !ok {
+		return nil, fmt.Errorf("could not find %s in map", key)
+	}
+
+	vb, ok := v.(*values.Bytes)
+	if !ok {
+		return nil, fmt.Errorf("value is not bytes: %s", key)
+	}
+
+	delete(m.Underlying, key)
+	return vb.Underlying, nil
+}
+
+func (c *Compute) executeWithModule(module *host.Module, config []byte, req capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	executeStart := time.Now()
+	capReq := capabilitiespb.CapabilityRequestToProto(req)
+
+	wasmReq := &wasmpb.Request{
+		Id:     uuid.New().String(),
+		Config: config,
+		Message: &wasmpb.Request_ComputeRequest{
+			ComputeRequest: &wasmpb.ComputeRequest{
+				Request: capReq,
+			},
+		},
+	}
+	resp, err := module.Run(wasmReq)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	cresppb := resp.GetComputeResponse().GetResponse()
+	if cresppb == nil {
+		return capabilities.CapabilityResponse{}, errors.New("got nil compute response")
+	}
+
+	cresp, err := capabilitiespb.CapabilityResponseFromProto(cresppb)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, errors.New("could not convert response proto into response")
+	}
+
+	computeWASMExec.WithLabelValues(
```

Comment:
is OTEL tracing a future use case here instead of measuring this execution time via prometheus?

Reply:
Yes I think so :)